### PR TITLE
Network error handling 

### PR DESCRIPTION
**Pull request by stafyniaksacha**
## Bugfixes
- SocketIO: handle network errors (like we do in Websocket protocol wrapper)
- Websocket: arguments can differs from NodeJS and Browsers
- Auto resubscribe when a reconnection occurs after a network error

### CODE REVIEW

File: src/networkWrapper/wrappers/socketio.js [hunk 1]
```diff
@@ -12,11 +22,75 @@ function SocketIO(host, port, ssl) {
...
+    var self = this;
+
     this.socket = window.io((this.ssl ? 'https://' : 'http://') + this.host + ':' + this.port, {
       reconnection: autoReconnect,
       reconnectionDelay: reconnectionDelay,
       forceNew: true
     });
+
+    this.socket.on('connect', function() {
+      if (self.wasConnected) {
```
Comment (benoitvidis) [inline src/networkWrapper/wrappers/socketio.js]: According to socket.io documentation, `connect` event is fired "upon a connection including a successful reconnection".
If this is the case, don't we have a double trigger with the `on('reconnect',..)` one?

File: src/networkWrapper/wrappers/socketio.js [hunk 2]
```diff
@@ -12,11 +22,75 @@ function SocketIO(host, port, ssl) {
...
+        handler(error);
+      });
+
+      if (autoReconnect && !self.retrying) {
+        self.retrying = true;
+        setTimeout(function () {
+          self.retrying = false;
+          self.connect(autoReconnect, reconnectionDelay);
+        }, reconnectionDelay);
+      }
```
Comment (benoitvidis) [inline src/networkWrapper/wrappers/socketio.js]: isn't this part supposed to be natively handled by socket.io `reconnect` options?

**Comment by ballinette** [comment]
missing description of what is the issue this PR fixes ?

File: src/networkWrapper/wrappers/socketio.js (hunk 1 above)
Comment (stafyniaksacha) [inline src/networkWrapper/wrappers/socketio.js]: Nope, this is different cases:
- reconnection on `connect` event is used when proxy close connection
- reconnection on `reconnect` event is used when socket hang out and come back again via socket io reconnection logic

File: src/networkWrapper/wrappers/socketio.js (hunk 2 above)
Comment (stafyniaksacha) [inline src/networkWrapper/wrappers/socketio.js]: I don't know why but reconnection from socketio does not work... (love socketio <3)

**Comment by stafyniaksacha** [comment]
@ballinette @benoitvidis sorry for the description, i wasn't saved it ^^

File: src/Room.js [hunk 3]
```diff
@@ -212,7 +213,16 @@ Room.prototype.renew = function (filters, notificationCB, cb) {
   self.kuzzle.subscriptions.pending[self.id] = self;
 
   subscribeQuery.body = self.filters;
-  subscribeQuery = self.kuzzle.addHeaders(subscribeQuery, this.headers);
+  subscribeQuery = self.kuzzle.addHeaders(subscribeQuery, self.headers);
+
+  networkListener = function() {
```
Comment (ballinette) [inline src/Room.js]: I don't like this behaviour (too many "addListener" and "removeListener").

Better just add following lines into `src/Kuzzle.js` (at line [365](https://github.com/kuzzleio/sdk-javascript/blob/321211f442a14dad8136a9ed4d3dee112905d63f/src/Kuzzle.js#L365)) :
```javascript
self.network.onConnectError(function (error) {
// (...)
  disableAllSubscriptions.call(self);
// (...)
});

function disableAllSubscriptions() {
  var self = this;

  Object.keys(self.subscriptions).forEach(function (roomId) {
    Object.keys(self.subscriptions[roomId]).forEach(function (subscriptionId) {
      var subscription = self.subscriptions[roomId][subscriptionId];
      subscription.subscribing = false;
    });
  });
}
```

It should work, but without a complex listener management.

Comment (stafyniaksacha) [inline src/Room.js]: Work like a charm !

File: src/networkWrapper/wrappers/socketio.js (hunk 1 above)
Comment (benoitvidis) [inline src/networkWrapper/wrappers/socketio.js]: Well, I see that you intend to use them differently but this is not what the doc says. `connect` is fired each time a connection is made, including socket.io reconnections.

File: src/networkWrapper/wrappers/socketio.js (hunk 2 above)
Comment (benoitvidis) [inline src/networkWrapper/wrappers/socketio.js]: Hum.. If this is the case, shouldn't we then just not use it at all? 
If it happens to work as expected, I still see an double trigger issue btw the connect and reconnect handlers (latest never being fired).

File: src/networkWrapper/wrappers/socketio.js (hunk 1 above)
Comment (stafyniaksacha) [inline src/networkWrapper/wrappers/socketio.js]: Ok, so I removed `reconnect` listener

**Pull request merged**